cmd/upgrade/app/v1alpha1/result.go

```diff
@@ -0,0 +1,258 @@
+	}
+	if len(urb.Tasks) == 0 {
+		validationErrs = append(validationErrs, errors.New("missing tasks"))
+	}
+	if len(validationErrs) != 0 {
+		urb.errors = append(urb.errors, validationErrs...)
+		return errors.Errorf("validation error(s) found: %v", validationErrs)
+	}
+	return nil
+}
+
+// Build builds a new instance of UpgradeResult with the
+// helps of UpgradeResultBuilder
+func (urb *UpgradeResultBuilder) Build() (res *upgrade.UpgradeResult, err error) {
+	err = urb.validate()
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+	}
+	l := jobNameLabelKey + "=" + urb.InstanceName +
+		"," + itemNameLabelKey + "=" + urb.ResourceDetails.Name +
+		"," + itemNamespaceLabelKey + "=" + urb.ResourceDetails.Namespace +
+		"," + itemKindLabelKey + "=" + urb.ResourceDetails.Kind
+	opts := metav1.ListOptions{
+		LabelSelector: l,
+	}
+	urList, err := upgraderesult.KubeClient(
+		upgraderesult.WithNamespace(urb.InstanceNamespace)).
+		List(opts)
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+	}
+	if len(urList.Items) == 1 {
+		return &urList.Items[0], nil
+	} else if len(urList.Items) == 0 {
+		ur, err := urb.getUpgradeResultObj()
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+		}
+
+		urb.UpgradeResult.object, err = upgraderesult.KubeClient(
+			upgraderesult.WithNamespace(urb.InstanceNamespace)).
+			Create(ur)
+		if err != nil {
+			return nil,
+				errors.Wrapf(err, "failed to build UpgradeResultCR: %s", urb)
+		}
+		return urb.UpgradeResult.object, nil
+	}
+	return nil,
+		errors.Errorf(`failed to build UpgradeResultCR:
+		multiple upgrade result cr found for resource: %v
+		upgrade result crs: %v`, urb.ResourceDetails, urList)
+}
+
+// getUpgradeResultObj returns UpgradeResult Object for given resource
+func (urb *UpgradeResultBuilder) getUpgradeResultObj() (*upgrade.UpgradeResult, error) {
+	tm, err := urb.getTypeMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	om, err := urb.getObjectMeta()
+	if err != nil {
+		return nil, err
+	}
+
+	return upgraderesult.NewBuilder().
+		WithTypeMeta(*tm).
+		WithObjectMeta(*om).
+		WithResultConfig(*urb.ResourceDetails, urb.UpgradeConfig.Data...).
+		WithTasks(urb.Tasks...).
+		Build()
+}
+
+// getTypeMeta returns metav1.TypeMeta for upgrade result cr
+func (urb *UpgradeResultBuilder) getTypeMeta() (tm *metav1.TypeMeta, err error) {
+	return typemeta.NewBuilder().
+		WithKind("UpgradeResult").
+		WithAPIVersion("openebs.io/v1alpha1").
+		Build()
+}
+
+// getTypeMeta returns metav1.ObjectMeta for upgrade result cr
+func (urb *UpgradeResultBuilder) getObjectMeta() (tm *metav1.ObjectMeta, err error) {
+	rand.Seed(time.Now().UnixNano())
+	letters := "abcdefghijklmnopqrstuvwxyz0123456789"
+	b := make([]byte, 5)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+	name := urb.UpgradeConfig.CASTemplate + "-" + string(b)
```

Comment:
Can the name be generated based on `job pod name`?
Can we make use of K8s suffix generation?
In other words, UpgradeResult name will be prefixed with this `instance/job pod name`. I guess pod name is autogenerated.

Reply:
yes we can use. updated.